Optimize Fractional Limit/Offset memory usage

0xgouda opened the pull request:
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Optimize performance and memory usage for fractional `LIMIT` and `OFFSET`.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

---

In https://github.com/ClickHouse/ClickHouse/pull/88755 `Fractional{Limit,Offset}` were added:

- For simplicity, it took the naive approach of caching the entire input data in memory, calculating the integral limit/offset, applying the normal limit/offset logic, and dropping the data from memory at the very end.
- Now we optimize that by smartly detecting blocks that will 100% be dropped or pushed to output and dropping/pushing them as early as possible while still in the pull phase.
- This achieves a huge memory usage reduction for `Fractional Offset` specifically, so if we have an `OFFSET 0.x` clause we achieve an approximate memory usage lower than `(1 - 0.x) * data_size` while in previous version the peak memory usage was always approximately equivalent to the `0.95 * data_size` all the time, e.g.:

    ```sql
    SELECT *
    FROM numbers(1000000000.)
    OFFSET 0.999999999

    Query id: 6ace27c4-2ebe-43bb-9a91-cc821e998282

       ┌────number─┐
    1. │ 999999999 │ -- 1000.00 million
       └───────────┘

    1 row in set. Elapsed: 0.238 sec. Processed 1.00 billion rows, 8.00 GB (4.21 billion rows/s., 33.68 GB/s.)
    Peak memory usage: 89.81 KiB. # instead of 7.45 GB previously
    ```
- For `LIMIT 0.n` alone without a companion `OFFSET 0.x` or `OFFSET x` the memory usage reductions is minimal, typically peak memory usage of 85% from total data_size must fit in-memory instead of 95% previously.
    - Because here the blocks are not dropped but rather passed to the next processor.
    - For smaller `0.n` numbers the memory usage reduction gets smaller and smaller.
    
- Noting that data is dropped/pushed in blocks, so all of this will depend on the block size.

0xgouda committed:
Optimize `FractionalOffsetTransform` memory usage.

0xgouda committed:
Optimize `FractionalLimitTransform` memory usage.

If `fractional_offset` specified, In pull phase: continuously drop entire
chunks that we became sure will be offseted,
don't cache all input and wait for push phase to filter them.

0xgouda committed:
More optimization for memory usage in `FractionalLimitTransform`

If no `offset_fraction` specified, we can push blocks while still
in pull phase, e.g., if we have 10 blocks with the same number of rows
and `limit_fraction` 0.1 we can safely push the first block to output.

0xgouda committed:
Improve comments in `FractionaOffsetTransform`

0xgouda committed:
Improve comments and var names in `FractionalLimitTransform`

0xgouda committed:
Add 2 spaces to trigger new line in `Fractional{Limit,Offset}` docs.

0xgouda committed:
Fix style errors

0xgouda commented:
@nihalzp I guess you would like to review this :).

0xgouda committed:
Fix line spacing in `offset.md` docs.

0xgouda committed:
Merge branch 'master' into optimise-fractional-offset

0xgouda committed:
Fix fractional limit docs, we round to `next` not `nearest`.

0xgouda commented:
ping @nihalzp

nihalzp reviewed (commented):
`OffsetTransform` looks good. For `LimitTransform` the following query is giving incorrect result:

```sql
SET max_block_size = 10;

SELECT number
FROM numbers(20)
ORDER BY number
LIMIT 0.5 OFFSET 5;
```

It returns only 5 rows. But should return 10 rows -
https://fiddle.clickhouse.com/b686e87a-87d9-4010-a93f-3d9a54aa0422

Please have a look.

I think we should have some tests where we use very small block size. Block size settings controls how many rows a chunk will contain. The existing tests could be replicated again with smaller block size to make sure behavior is preserved.

0xgouda committed:
Fix fractional limit early pushing logic.

0xgouda committed:
Test early pushing/eviction in fractional limit/offset.

0xgouda committed:
Improve comments.

0xgouda commented:
Ummm, The with ties test is failing, give me some time to debug it.

0xgouda committed:
Make `03752_fractional_limit_offset_small_blocks` test faster.

0xgouda committed:
Fix early pushing logic.

0xgouda committed:
Improve comment.

0xgouda committed:
Fix overflow in `remaining_offset` calculation.

0xgouda commented:
@nihalzp 
All should be good now; failing tests are unrelated. Please re-review.

0xgouda committed:
Fix Integral offset's chunk ignore logic.

Use `rows_cnt <= offset` instead of `rows <= remaining_offset`
because if chunk size > offset and last chunks has
number of rows < offset, it will be ignored, which is wrong.

0xgouda committed:
duplicate tests but using block size 3.

0xgouda committed:
Strengthen test.

0xgouda commented:
@nihalzp 
Everything should be good by now. Please re-review.

nihalzp reviewed (commented):
Previous problem seems to be resolved!

I think there might be an issue with `WITH TIES`. For example,

```sql
SELECT
    count() AS rows,
    min(number) AS min_n,
    max(number) AS max_n,
    arraySort(groupArray(number)) AS nums
FROM
(
    SELECT number
    FROM numbers(50)
    ORDER BY number
    LIMIT 0.3 WITH TIES
    SETTINGS
        max_block_size = 10,
        max_threads = 1
);
```
should return count of 15 rows; but in the PR, it gives count of only 10 rows. https://fiddle.clickhouse.com/faabe077-e7e8-44a9-9339-2f8b72710c94

0xgouda committed:
Fix `WITH TIES` logic

Fixes the `previous_row_chunk` assignment logic to ensure
that it doesn't have a value if we still haven't finished
our normal `limit + offset` processing.

That avoids making `splitChunk` break our logic as it expects
a value for `previous_row_chunk` only if we are finished with
output-ing `limit + offset`.

0xgouda committed:
Fix Styling

0xgouda committed:
duplicate small blocks test but with using `WITH TIES`.

The data doesn't contain any ties, so the operator mustn't
show any effect.

nihalzp commented on src/Processors/FractionalLimitTransform.cpp:
```diff
@@ -172,6 +185,66 @@ FractionalLimitTransform::Status FractionalLimitTransform::pullData(PortsData &
+    /// will 100% get pushed and push them as early as possible.
+    if (offset_fraction == 0 && !chunks_cache.empty())
+    {
+        auto output = *chunks_cache.front().output_port;
```
```suggestion
        auto & output = *chunks_cache.front().output_port;
```

nihalzp commented on src/Processors/FractionalLimitTransform.cpp:
```diff
@@ -172,6 +185,66 @@ FractionalLimitTransform::Status FractionalLimitTransform::pullData(PortsData &
+        if (rows_read_from_cache < offset)
+            remaining_offset = offset - rows_read_from_cache;
+
+        auto curr_limit = std::ceil(rows_cnt * limit_fraction);
```
```suggestion
        UInt64 curr_limit = std::ceil(rows_cnt * limit_fraction);
```
Otherwise, `cur_limit` would have type `double` which could potentially create problems later.

nihalzp reviewed (commented):
The previous issues are resolved!

I have left few comments.

Additionally, even though we enforce invariants such as both fractional offset and integral offset cannot be nonzero and fractional offset need to be in `(0, 1)` if nonzero in the planner. I think it might be good practice to enforce these invariants in the constructor because in the future, this transform might be called from other places. We can throw `LOGICAL ERROR` exception otherwise if the invariants are not met.

Our tests mostly uses `ORDER BY` in the query which makes, if I am not wrong, `num_streams = 1` and we do not test many of the branches that happen when `num_streams > 1`. To make it convenient, we can replicate our tests and remove `ORDER BY`. We can do a weak test result validation in the following form:

```sql
SELECT count() FROM ( [the existing query without ORDER BY]);
```

0xgouda committed:
Don't close input on output finish

It may belong to other pairs.

Also as much chunks that belong to this output.

0xgouda committed:
Cast all `ceil()` returns to `UInt64`

0xgouda committed:
throw logical error in constructor on fraction >= 1 or <= 0

0xgouda committed:
Fix overflow caused by subtraction on UInt64

0xgouda committed:
Improve fractional limit/offset out of range exception message

0xgouda committed:
Fix `executeOffset()` in old planner

It used fractional offset for -0, 0 values but now
it doesn't add an offset step.

0xgouda committed:
duplicate fractional limit tests but with multiple streams

0xgouda committed:
Make test faster

0xgouda committed:
Pop only once from cache on closed output in pull phase

No need to use a while loop of poping from front there is no
guarantee that the next chunk will have the same output

0xgouda commented:
> @0xgouda thanks for PR, please add a performance test

Hi @fm4v, thanks for pointing this out. I still have some improvements to make regarding memory usage. I will add the test with them in a future PR soon.
